### ISSUE TITLE
Lint/ShadowingOuterLocalVariable should not be triggered in {class|module|instance}_exec

### ISSUE THREAD

**Issue by lordofthelake** [rubocop-hq/rubocop#3692]
This code triggers a `Lint/ShadowingOuterLocalVariable` offence:

```ruby
def define_module(foo, bar)
  anonymous_module = Module.new
  anonymous_module.class_exec(foo, bar) do |foo, bar|
    define_method foo { bar }
  end
end
```

Since the blocks passed to `class_exec`/`module_exec`/`instance_exec` don't act as a closure (and the arguments are specifically there to be "pass-through"), IMHO the warning should not be triggered, as there is no "shadowing" going on (the variables out of the block wouldn't be available anyway).

## RuboCop version

```
$ rubocop -V
0.43.0 (using Parser 2.3.1.4, running on ruby 2.3.1 x86_64-darwin15)
```

**Comment by mikegee** [comment]
The example code also produces warnings from Ruby:

```
% ruby -cw test/test.rb
test/test.rb:3: warning: shadowing outer local variable - foo
test/test.rb:3: warning: shadowing outer local variable - bar
Syntax OK
```

So I suppose this issue should be raised with Ruby itself. ¯\\\_(ツ)_/¯

**Comment by jonas054** [comment]
> Since the blocks passed to class_exec/module_exec/instance_exec don't act as a closure (and the arguments are specifically there to be "pass-through"), IMHO the warning should not be triggered, as there is no "shadowing" going on (the variables out of the block wouldn't be available anyway).

I have not found this to be true. Consider this:

```ruby
def define_module(foo, bar)
  anonymous_module = Module.new
  anonymous_module.class_exec(foo, bar+1) do |foo2, bar2|
    define_method(foo2) { puts "#{bar}, #{bar2}" }
  end
  anonymous_module
end

Mod = define_module(:echo, 3)

class C
  include Mod
end

C.new.echo # prints "3, 4"
```
Both `bar` and `bar2` are available inside the block. If `bar2` is renamed `bar` it does shadow the outer `bar`.

**Comment by avtrujillo** [comment]
A similar bug happens here: https://pastebin.com/xkrWJdQi

animals = ["dog", "cat", "chicken", "shark"]
 
if 2 + 2 == 5 #obviously this condition will never be true
  # at this point, animal is still undefined
  animal = "frog"
  puts animal
else
  # animal is now defined as nil, rather than
  # being undefined or defined as "frog"
  puts(animals.find { |animal| animal.length == 5})
  # Rubocop will cite me here for shadowing, becuase it considers animal to
  # still be defined.
end
 
# to see only the relevant offense, run the line below in a terminal:
# rubocop Test.rb --only Lint

**Comment by avtrujillo** [comment]
A similar problem occurs here: https://pastebin.com/xkrWJdQi

`animals = ["dog", "cat", "chicken", "shark"]
 
if 2 + 2 == 5 #obviously this condition will never be true
  # at this point, animal is still undefined
  animal = "frog"
  puts animal
else
  # animal is now defined as nil, rather than
  # being undefined or defined as "frog"
  puts(animals.find { |animal| animal.length == 5})
  # Rubocop will cite me here for shadowing, becuase it considers animal to
  # still be defined.
end
 
# to see only the relevant offense, run the line below in a terminal:
# rubocop Test.rb --only Lint`

**Comment by bbatsov** [comment]
There hasn't been much activity on this ticket and our Core Team is spread too thin on so many tasks, so I'll just close it for the sake of having a cleaner lists of tasks to focus on. We'd gladly review a PR, but it's unlikely that'd we're going to tackle this ourselves in the foreseeable future.
